src/main/kotlin/com/sourcegraph/cody/context/EnterpriseEnhancedContextStateController.kt

```diff
@@ -0,0 +1,343 @@
+ *   be read and written, however the agent may add a repository it has picked up and included
+ *   automatically by examining the project.
+ * - The chat sidebar UI presents a view of enhanced context to the user. (Including a text field in
+ *   a popup, however that is only *read* by the controller so does not appear here--see
+ *   [EnterpriseEnhancedContextStateController.updateRawSpec].)
+ */
+interface ChatEnhancedContextStateProvider {
+  /** Updates JetBrains Cody's "chat history" copy of enhanced context state. */
+  fun updateSavedState(updater: (EnhancedContextState) -> Unit)
+
+  /** Updates the Agent-side state for the chat. */
+  fun updateAgentState(repos: List<Repo>)
+
+  /** Pushes a UI update to the chat side panel. */
+  fun updateUI(repos: List<RemoteRepo>)
+
+  /** Displays a message that remote repository resolution failed. */
+  fun notifyRemoteRepoResolutionFailed()
+
+  /** Displays a message that the user has reached the maximum number of remote repositories. */
+  fun notifyRemoteRepoLimit()
+}
+
+/**
+ * Reconciles the multiple, asynchronously updated copies of enhanced context state.
+ *
+ * Changes follow this flow:
+ * 1. A chat is restored ([loadFromChatState]) which synthesizes a [rawSpec] and a
+ *    `model.manuallyDeselected` set.
+ * 2. When the raw spec is updated, we parse it and produce a "speculative set of repos"
+ *    ([updateSpeculativeRepos]) These have not been resolved by the backend and may be totally
+ *    bogus.
+ * 3. When the speculative repos are resolved ([onResolvedRepos]) we can filter the
+ *    `model.manuallyDeselected` ones and request the Agent to focus on a set of repositories
+ *    (`chat.updateAgentState`).
+ * 4. When the Agent has updated its state ([onAgentStateUpdated]) we finally learn which
+ *    repositories are actually used, whether a repository is implicitly included by the Agent based
+ *    on the project, and whether a repository is filtered by Context Filters.
+ * 5. Finally, we can [updateUI].
+ *
+ * When the user updates the raw spec, the same process happens from step 2 to step 4, however we
+ * also `chat.updateSavedState` to save the changes to the JetBrains-side copy of chat history.
+ *
+ * When the user checks and unchecks repositories, we already have all the resolved repository
+ * details. We just update the JetBrains-side copy of chat history (`chat.updateSavedState`) and do
+ * the `chat.updateAgentState` -> [onAgentStateUpdated] flow.
+ */
+class EnterpriseEnhancedContextStateController(
+    val project: Project,
+    val chat: ChatEnhancedContextStateProvider
+) {
+  private val logger = Logger.getInstance(EnterpriseEnhancedContextStateController::class.java)
+  private val model_ = EnterpriseEnhancedContextModel()
+  private var epoch = 0
+
+  val rawSpec: String
+    get(): String = model_.rawSpec
+
+  private fun <T> withModel(f: (EnterpriseEnhancedContextModel) -> T): T {
+    assert(!ApplicationManager.getApplication().isDispatchThread) {
```

Comment:
Maybe nicer to do similar to the `@RequiresEdt` function annotations we have

Reply:
Those annotations do nothing, so I'm not using them.

When the annotations support Kotlin and actually do something, we can replace these assertions with annotations.

Reply:
That is not strictly true @dominiccooney . They do whave effect, but I think only for the newer IntelliJ versions:

https://github.com/JetBrains/intellij-community/blob/master/platform/core-api/src/com/intellij/util/concurrency/annotations/RequiresEdt.java


```
/**
 * Methods and constructors annotated with {@code RequiresEdt} must be called from the Event Dispatch Thread only.
 * Parameters annotated with {@code RequiresEdt} must be callables and are guaranteed to be called from the Event Dispatch Thread.
 * <p/>
 * Aside from a documentation purpose, the annotation is processed by the {@link org.jetbrains.jps.devkit.threadingModelHelper}.
 * The plugin instruments annotated elements with {@link Application#assertIsDispatchThread()} calls
 * to ensure annotation's contract is not violated at runtime. The instrumentation can be disabled
 * by setting {@link RequiresEdt#generateAssertion()} to {@code false}.
 * <p/>
 * <b>Important:</b> the instrumentation has limitations. Please read the docs
 * of the {@link org.jetbrains.jps.devkit.threadingModelHelper} to learn about them.
 *
 * @see <a href="https://plugins.jetbrains.com/docs/intellij/general-threading-rules.html">General Threading Rules</a>
 * @see Application#assertIsDispatchThread()
 * @see Application#invokeLater(Runnable, ModalityState)
 * @see Application#invokeAndWait(Runnable, ModalityState)
 */
```